runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -906,78 +906,195 @@ allocateWriteBarrierInternalPointerRegister(TR::CodeGenerator * cg, TR::Node * s
+   generateS390LabelInstruction(cg, TR::InstOpCode::label, node, equalRegion);
+   if (node->getOpCode().isMax())
+      {
+      // For Max calls, checking if first operand is +0, then we are done, otherwise fall through for swap
+      generateRXEInstruction(cg, tdcOpcode, node, operand1, generateS390MemoryReference(0x800, cg), 0);  // operand1 is +0 ?
+      generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_MASK4, node, cFlowRegionEnd);   // it is +0
+      }
+   else if (node->getOpCode().isMin())
+      {
+      // For Min calls, checking if first operand is not +0, then we are done, otherwise fall through for swap
+      generateRXEInstruction(cg, tdcOpcode, node, operand1, generateS390MemoryReference(0x400, cg), 0);  // operand1 is -0 ?
+      generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_MASK4, node, cFlowRegionEnd);
+      }
+
+   generateS390LabelInstruction(cg, TR::InstOpCode::label, node, swap);
+   // Move resulting operand to operand1 as fallthrough for alternate Condition Code
+   TR::InstOpCode::Mnemonic moveRROp = node->getOpCode().isDouble() ? TR::InstOpCode::LDR : TR::InstOpCode::LER;
+   generateRREInstruction(cg, moveRROp, node, operand1, operand2);
+
+   TR::RegisterDependencyConditions* deps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 2, cg);
+   deps->addPostConditionIfNotAlreadyInserted(operand1, TR::RealRegister::AssignAny);
+   deps->addPostConditionIfNotAlreadyInserted(operand2, TR::RealRegister::AssignAny);
+
+   generateS390LabelInstruction(cg, TR::InstOpCode::label, node, cFlowRegionEnd, deps);
+   cFlowRegionEnd->setEndInternalControlFlow();
+
+   node->setRegister(operand1);
+   cg->decReferenceCount(lhsNode);
+   cg->decReferenceCount(rhsNode);
+
+   return operand1;
+   }
+
+TR::Register*
+J9::Z::TreeEvaluator::fminEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   if (cg->getSupportsVectorRegisters())
+      {
+      return fpMinMaxVectorHelper(node, cg, TR::DataTypes::Float, false);
+      }
+   return fpMinMaxHelper(node, cg, TR::DataTypes::Float, false);
+   }
+
+TR::Register*
+J9::Z::TreeEvaluator::dminEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   cg->generateDebugCounter("z13/simd/doubleMin", 1, TR::DebugCounter::Free);
+   if (cg->getSupportsVectorRegisters())
+      {
+      return fpMinMaxVectorHelper(node, cg, TR::DataTypes::Double, false);
+      }
+   return fpMinMaxHelper(node, cg, TR::DataTypes::Double, false);
+   }
+
+TR::Register*
+J9::Z::TreeEvaluator::fmaxEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   if (cg->getSupportsVectorRegisters())
+      {
+      return fpMinMaxVectorHelper(node, cg, TR::DataTypes::Float, true);
+      }
+   return fpMinMaxHelper(node, cg, TR::DataTypes::Float, true);
+   }
+
+TR::Register*
+J9::Z::TreeEvaluator::dmaxEvaluator(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   cg->generateDebugCounter("z13/simd/doubleMax", 1, TR::DebugCounter::Free);
```

Comment:
Debug counter is for SIMD acceleration - So should be only generated when we generate fpMinMaxVector implementation - So this should be in the if condition - Similar thing in all dmin/dmax/fmin/fmax cases.